File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/study/StudyGrammarGenerator.kt
package org.wyrdsekai.app.engine.study

import org.wyrdsekai.app.protocol.Exit
import org.wyrdsekai.app.protocol.RoomObject

/**
 * Generates GBNF grammars for the Study room on phone.
 *
 * Kotlin port of common/grammar/RoomGrammarGenerator.generateStudy().
 * Constrains 0.6B model output to valid Study actions.
 * Speech is always free. Actions constrained to room state.
 */
object StudyGrammarGenerator {

    /**
     * Generate a GBNF grammar for the Study room.
     * @param exits Available exits from the current room
     * @return GBNF grammar string
     */
    fun generate(exits: List<Exit> = emptyList()): String {
        val rules = mutableListOf<String>()
        val alts = mutableListOf<String>()

        // Free speech — always available, unconstrained
        rules.add("speech ::= \"say:\" freetext")
        rules.add("freetext ::= [^\\n]+")
        alts.add("speech")

        // Emote
        rules.add("emote ::= \"emote:\" freetext")
        alts.add("emote")

        // Navigation
        if (exits.isNotEmpty()) {
            val dirs = exits.joinToString(" | ") { "\"${escapeGbnf(it.direction)}\"" }
            rules.add("navigate ::= \"go:\" direction")
            rules.add("direction ::= $dirs")
            alts.add("navigate")
        }

        // Study objects (fixed set)
        rules.add("use_object ::= \"use:\" object_name")
        rules.add("object_name ::= \"journal\" | \"desk\" | \"shelves\" | \"pinboard\"")
        alts.add("use_object")

        // Look
        rules.add("look ::= \"look\" | \"look:\" freetext")
        alts.add("look")

        // Study-specific verbs
        val studyVerbs = listOf(
            "journal_write" to true,
            "journal_search" to true,
            "journal_private" to true,
            "note_add" to true,
            "note_search" to true,
            "pin" to true,
            "remind" to true,
            "summarize" to false,
            "digest" to false,
        )
        for ((name, hasArg) in studyVerbs) {
            if (hasArg) {
                rules.add("$name ::= \"$name:\" freetext")
            } else {
                rules.add("$name ::= \"$name\"")
            }
            alts.add(name)
        }

        // Root: pick one action
        val root = "root ::= ${alts.joinToString(" | ")}"
        return root + "\n\n" + rules.joinToString("\n") + "\n"
    }

    private fun escapeGbnf(s: String): String =
        s.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
}

File: clients/rn/src/engine/study/StudyGrammarGenerator.ts
/**
 * Generates GBNF grammars for the Study room on phone.
 * TypeScript port of KMP's StudyGrammarGenerator.kt.
 *
 * Constrains 0.6B model output to valid Study actions.
 * Speech is always free. Actions constrained to room state.
 */

import type { Exit } from '../../protocol/models';

function escapeGbnf(s: string): string {
  return s.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
}

/**
 * Generate a GBNF grammar for the Study room.
 * @param exits Available exits from the current room
 * @returns GBNF grammar string
 */
export function generateStudyGrammar(exits: Exit[] = []): string {
  const rules: string[] = [];
  const alts: string[] = [];

  // Free speech — always available, unconstrained
  rules.push('speech ::= "say:" freetext');
  rules.push('freetext ::= [^\\n]+');
  alts.push('speech');

  // Emote
  rules.push('emote ::= "emote:" freetext');
  alts.push('emote');

  // Navigation
  if (exits.length > 0) {
    const dirs = exits.map(e => `"${escapeGbnf(e.direction)}"`).join(' | ');
    rules.push('navigate ::= "go:" direction');
    rules.push(`direction ::= ${dirs}`);
    alts.push('navigate');
  }

  // Study objects (fixed set)
  rules.push('use_object ::= "use:" object_name');
  rules.push('object_name ::= "journal" | "desk" | "shelves" | "pinboard"');
  alts.push('use_object');

  // Look
  rules.push('look ::= "look" | "look:" freetext');
  alts.push('look');

  // Study-specific verbs
  const studyVerbs: [string, boolean][] = [
    ['journal_write', true],
    ['journal_search', true],
    ['journal_private', true],
    ['note_add', true],
    ['note_search', true],
    ['pin', true],
    ['remind', true],
    ['summarize', false],
    ['digest', false],
  ];

  for (const [name, hasArg] of studyVerbs) {
    if (hasArg) {
      rules.push(`${name} ::= "${name}:" freetext`);
    } else {
      rules.push(`${name} ::= "${name}"`);
    }
    alts.push(name);
  }

  const root = `root ::= ${alts.join(' | ')}`;
  return root + '\n\n' + rules.join('\n') + '\n';
}
